### PR TITLE
fix: Do not make a `FastCount` for csv if `pre_slice` is set

### LINKED ISSUE #27534: pl.len() returns full size on a slized LazyFrame
### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

df = pl.scan_csv(b"""a\n1\n2\n3\n4""").slice(0, 2)
length = int(df.select(pl.len()).collect().item())
print(f"{length = }")
length = int(df.select(pl.len()).collect(optimizations=pl.QueryOptFlags.none()).item())
print(f"{length = }")
```

### Log output

```shell
_init_credential_provider_builder(): credential_provider_init = None
async thread count: 32
blocking thread count: 512
polars-stream: updating graph state
updating io-sink[single-file[csv]], before: [Blocked] []
updating io-sink[single-file[csv]], after: [Ready] [] (took 90ns)
updating in-memory-source, before: [] [Ready]
updating in-memory-source, after: [] [Ready] (took 269ns)
updating io-sink[single-file[csv]], before: [Ready] []
updating io-sink[single-file[csv]], after: [Ready] [] (took 63ns)
polars-stream: running in-memory-source in subgraph
polars-stream: running io-sink[single-file[csv]] in subgraph
upload_chunk_size: 33554432
total memory: 30.292 GiB (32525324288 bytes)
upload_concurrency: 30
[CsvWriterStarter]: base_allocation_size: 640000
io-sink[single-file[csv]]: start_single_file_sink_pipeline: file_writer_starter: csv, takeable_rows_provider: TakeableRowsProvider { max_size: NonZeroRowCountAndSize { num_rows: 25600, num_bytes: 8388608 }, byte_size_min_rows: 256, allow_non_max_size: true }, inflight_morsel_limit: 36, upload_chunk_size: 33554432, upload_concurrency: 30, io_metrics: false, build_sinked_path_info_list: false
polars-stream: done running graph phase
polars-stream: updating graph state
updating io-sink[single-file[csv]], before: [Ready] []
updating io-sink[single-file[csv]], after: [Ready] [] (took 101ns)
updating in-memory-source, before: [] [Ready]
updating in-memory-source, after: [] [Done] (took 330ns)
updating io-sink[single-file[csv]], before: [Done] []
io-sink[single-file[csv]]: Join on task_handle (recv PortState::Done)
io-sink[single-file[csv]]: Statistics: total_size: RowCountAndSize { num_rows: 4, num_bytes: 32 }
updating io-sink[single-file[csv]], after: [Done] [] (took 1.014644ms)
updating in-memory-source, before: [] [Done]
updating in-memory-source, after: [] [Done] (took 160ns)
_init_credential_provider_builder(): credential_provider_init = None
sourcing csv scan file schema from: 'data.csv'
4
polars-stream: updating graph state
updating in-memory-sink, before: [Blocked] []
updating in-memory-sink, after: [Ready] [] (took 150ns)
updating multi-scan[csv], before: [] [Ready]
updating multi-scan[csv], after: [] [Ready] (took 32.646µs)
updating in-memory-sink, before: [Ready] []
updating in-memory-sink, after: [Ready] [] (took 52ns)
polars-stream: running multi-scan[csv] in subgraph
polars-stream: running in-memory-sink in subgraph
[CsvReaderBuilder]: prefetch_limit: 64
[MultiScanTaskInit]: 1 sources, reader name: csv, ReaderCapabilities(PRE_SLICE), n_readers_pre_init: 1, max_concurrent_scans: 1, disable_morsel_split: true
[MultiScanTaskInit]: predicate: None, deletion_files: None, skip files mask: None, predicate to reader: None
[MultiScanTaskInit]: scan_source_idx: 0, extra_ops: ExtraOperations { row_index: None, row_index_col_idx: 18446744073709551615, pre_slice: None, include_file_paths: None, file_path_col_idx: 18446744073709551615, predicate: None }
[MultiScanTaskInit]: Readers init: 1 / (1 total) (range: 0..1, filtered out: 0)
[MultiScan]: Initialize source 0
[ReaderStarter]: scan_source_idx: 0
[ReaderStarter]: max_concurrent_scans is 1, waiting..
[AttachReaderToBridge]: received reader (n_readers_received: 1)
[ReaderStarter]: scan_source_idx: 0: pre_slice_to_reader: None, external_filter_mask: None, file_iceberg_schema: None
[CsvFileReader]: project: 1 / 1, slice: None, use_async_prefetch: false
[CsvSource]: Start line splitting direct
start_reader_impl: scan_source_idx: 0, first_morsel_position: RowCounter { physical_rows: 0, deleted_rows: 0 }
start_reader_impl: scan_source_idx: 0, ApplyExtraOps::Noop, first_morsel_position: RowCounter { physical_rows: 0, deleted_rows: 0 }
[ReaderStarter]: Stopping (no more readers)
[MultiScanState]: Readers disconnected
polars-stream: done running graph phase
polars-stream: updating graph state
updating in-memory-sink, before: [Ready] []
updating in-memory-sink, after: [Ready] [] (took 53ns)
updating multi-scan[csv], before: [] [Ready]
updating multi-scan[csv], after: [] [Done] (took 52.043µs)
updating in-memory-sink, before: [Done] []
updating in-memory-sink, after: [Done] [] (took 82ns)
updating multi-scan[csv], before: [] [Done]
updating multi-scan[csv], after: [] [Done] (took 165ns)
2
```

### Issue description

When a csv is loaded with scan_csv, if taking a slice of it, `df.select(pl.len())` returns the full CSV length instead of the slice length. Disabling optimizations, it returns the correct size.

### Expected behavior

It should return the slice size

### Installed versions

<details>

```
Polars:              1.40.1
Index type:          UInt32
Platform:            Linux-6.17.0-5-generic-x86_64-with-glibc2.42
Python:              3.14.4 (main, Apr  9 2026, 10:35:55) [GCC 15.2.0]
Runtime:             rt32

----Optional dependencies----
Azure CLI            <not installed>
adbc_driver_manager  <not installed>
altair               <not installed>
azure.identity       <not installed>
boto3                <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               <not installed>
gevent               <not installed>
google.auth          <not installed>
great_tables         <not installed>
matplotlib           3.10.8
numpy                2.4.4
openpyxl             <not installed>
pandas               3.0.2
polars_cloud         <not installed>
pyarrow              24.0.0
pydantic             2.13.3
pyiceberg            <not installed>
sqlalchemy           <not installed>
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>

```

</details>

### PR DESCRIPTION
**Pull request by dsprenkels**
Fixes #27534

:robot: AI located the cause of the bug. I did the rest.

### CODE REVIEW

**Comment by coastalwhite** [comment]
Maybe as a future improvement. A `FastCount` with a `pre_slice` could be turned into `FastCount.min(offset + length).saturating_sub(offset)`.

File: py-polars/tests/unit/lazyframe/test_projections.py [hunk 1]
```diff
@@ -883,3 +868,35 @@ def test_projection_pushdown_filter_len_to_sum() -> None:
...
+
+@pytest.mark.parametrize(
+    ("sink", "scan"),
+    [
+        (pl.DataFrame.write_csv, pl.scan_csv),
+        (pl.DataFrame.write_parquet, pl.scan_parquet),
+        (pl.DataFrame.write_ipc, pl.scan_ipc),
+    ],
+)
+@pytest.mark.parametrize("slice", [None, (0, 5)])
```
Comment (nameexhaustion) [inline py-polars/tests/unit/lazyframe/test_projections.py]: Could we also test positive / negative offset slices, both overlapping and if they exceed the data length.

**Review by nameexhaustion — APPROVED** [review]
